### ISSUE TITLE
TimeFormat::checkTime problem with microseconds

### ISSUE THREAD

**Issue by hubipe** [consistence/consistence#22]
Hello,
I came across incompatibility with TimeFormat::checkTime method and microseconds. When creating new datetime from string containing microseconds using this:

```php
TimeFormat::createDateTimeImmutableFromFormat('Y-m-d\TH:i:s.u', '2017-10-26T16:48:39.139')
```

Exception is thrown:
```
Consistence\Time\TimeDoesNotMatchFormatException
Time "2017-10-26T16:48:39.139" does not match format "Y-m-d\TH:i:s.u"
```

Problem is in inconsistency between parameters available for PHP methods DateTime::createFromFormat and date. Date function has parameter "v" standing for miliseconds (3 fraction positions) and "u" (6 fraction positions). But DateTime::createFromFormat knows only "u". Therefore when calling createFromFormat with datetime string, "v" format parameter can't be used (exception is thrown here: https://github.com/consistence/consistence/blob/master/src/Time/TimeFormat.php#L160 and when "u" parameter is used, exception is thrown here: https://github.com/consistence/consistence/blob/master/src/Time/TimeFormat.php#L208

In my opinion dates 2017-10-26T16:48:39.139 and 2017-10-26T16:48:39.139000 are equivalent and exception should not be raised.

**Comment by VasekPurchart** [comment]
Hi, I know about the issue with `v` ever since I implemented these and even have a card about it in my internal Trello board. Unfortunately I still don't know how to handle this case correctly.

Using `v` is quite problematic due to the disparity you mentioned (fine when serializing, not recognized when parsed), that is why I have even withheld from listing it in the `TimeFormat` constants, unlike all the others.

> In my opinion dates 2017-10-26T16:48:39.139 and 2017-10-26T16:48:39.139000 are equivalent and exception should not be raised.

I don't agree with this handling of `u`, because that is exactly what is strict DateTime handling trying to prevent. The same could be said about `2017-1-1` being essentially the same as `2017-01-01`, but in the strict sense it is not. Because then this would not hold true:

```php
$timeString = '2017-10-26T16:48:39.139';
$format = 'Y-m-d\TH:i:s.u';
$parsedTime = TimeFormat::createDateTimeImmutableFromFormat($format, $timeString);
$timeString === $parsedTime->format($format); // false
```

The correct solution to this problem would be to parse `v` as intended. I hoped that this would be eventually implemented, because `v` was introduced only in `7.0`.  I tried to search for other people dealing with this disparity, but found nothing in the [bugtracker](https://bugs.php.net/search.php?cmd=display&search_for=millisecond&x=0&y=0), maybe you can try a few searches as well. Maybe then we can try filling a bug there.

Other way of dealing with this in the meantime would be:

1) Implementing parsing of `v` directly in Consistence
2) "Converting" the time to `u` and then using `u`
3) Not using `v` anywhere
4) Not use strict parsing for this

Writing the strict checking was really difficult due to almost all PHP functions carrying some kind of magic and not presenting react-able errors back while at the same time trying to not do any of the actual parsing differently in the non-magic use scenarios. This almost completely eliminates `1)`, because even if we had some kind of internal conversion of millisecond values to microseconds then this would be already bahaving differently than the other PHP methods, barring the fact, that I think this would be extremely difficult without dragging a lot of the actual format parsing in the code, for it to be working correctly.

This leaves us with more workaround methods, depending whether you are dealing with a particular case (small scale) or a general parsing. If you are dealing with a particular case and you **know** how the data is produced, you could append `000` and then use strict `u` parsing (`2)`), meaning that if something should go awry, the strict parsing would save the day.

Of course if you have control over the creation of the parsed data, much easier would be not to use `v` (`3)`), which is what I decided to do, regarding the disparity in PHP handling. But of course this is possible every time if you either have no control over the data or need this particular presision due to some other limitations.

Or you can simply not use strict parsing for this particular case (`4`), because the format in fact is not strict. You could try to come up with some kind of a combination of non-strict and strict - but in the end everything would be suboptimal, because you are effectively either already parsing it (when trying to evaluate only part of it non-strictly - which is similar to `1)`), or you are giving up strictness in the other parts of the string and you cannot be sure about the data then.

**Comment by hubipe** [comment]
Thank you for this broad explanation of the problem and suggested solutions. The easiest is (and that is actually how I handled it) is to give up strict parsing (4). Date comes from SOAP webservice, in which field is of type XMLSchema datetime. By definition, XMLSchema datetime don't use fixed digits in second fraction. In fact, in documentation (https://www.w3.org/TR/xmlschema-2/#dateTime-canonical-representation) it states 

> The fractional second string, if present, must not end in '0'

therefore I can't rely on same string representation coming from the service. Given that though, I can be almost certain the PHP parser will parse the string correctly.

**About (in)equality of 2017-10-26T16:48:39.139 and 2017-10-26T16:48:39.139000**

I still don't agree with you on this. Given your example, you stated, that modifiers for parsing the date has to be the same as for formatting. In my opinion, parsing method parameters are NOT the same as the formatting ones, because the "u" parameter in `createFromFormat `method correctly parses second fractions _up to_ 6 digits, but format parameter "u" gives you _always_ 6 digits. Although these letters are the same, their meaning is not.

Also `createFromFormat` accepts modifiers `#, ?, *, !, |` and `+`, which format method does not recognize. I know we're talking about strict parsing, but shouldn't you be able to parse this?

```php
DateTime::createFromFormat('Y#m#d', '2017-10-26');
```

That is 100% correct format, but Consistence would raise an exception about it as invalid format similarly as with the "u" parameter.

My point is that string comparison of input and string representation of parsed `DateTime` works only in a subset of all valid date time formats. Maybe this behavior is good in `TimeFormat::checkTime` method, but is string comparison desirable in `createFromFormat` method? Shouldn't this method just check ability to parse string and fact that given date string doesn't contain invalid dates such as 2017-11-31 or 25:00:00?

**Comment by VasekPurchart** [comment]
> I still don't agree with you on this. Given your example, you stated, that modifiers for parsing the date has to be the same as for formatting. In my opinion, parsing method parameters are NOT the same as the formatting ones, because the "u" parameter in `createFromFormat `method correctly parses second fractions _up to_ 6 digits, but format parameter "u" gives you _always_ 6 digits. Although these letters are the same, their meaning is not.

Yes, I know that the parameters have different meaning in parsing and formatting and that is exactly what is often misleading. This is the same situation as with the hours - when formatting `H` will always produce `01`, but when parsing both `1` and `01` are ok, making it ambiguous.

> My point is that string comparison of input and string representation of parsed `DateTime` works only in a subset of all valid date time formats. Maybe this behavior is good in `TimeFormat::checkTime` method, but is string comparison desirable in `createFromFormat` method? Shouldn't this method just check ability to parse string and fact that given date string doesn't contain invalid dates such as 2017-11-31 or 25:00:00?

Well both the `checkTime` and `createFromFormat` must definitely behave consistently, you can't have a `createFromFormat` result which would then fail in `checkTime`.

The problem with allowing more relaxed parsing was, if I remember correctly, that due to PHP's magic parsing nature it lead to even more errors. Which makes sense, because the less strict you are, the more wiggle room there is left for interpretation. This with combination that PHP does not really tell you well what happened during the parsing was why this lead me to thinking of the strict formats as first line of defense.

> Also createFromFormat accepts modifiers #, ?, *, !, | and +, which format method does not recognize. I know we're talking about strict parsing, but shouldn't you be able to parse this?

Thanks, this is actually a really good point and something I probably didn't really think of, when implementing this, since I was focused more on the formatting parameters. I certainly agree that the method should be able to parse these, but I will have to think about it, if it will be possible while preserving the other constraints (both design and implementation ones).

**Comment by hubipe** [comment]
PHP 7.3 added the parameter `v` to the [`createFromFormat`](https://www.php.net/manual/en/datetime.createfromformat.php) method, so the original issue is not an issue anymore, I suppose only the issues with the modifiers still persist, but that may be rather different issue for someone. So maybe we could close this issue?
